README.md
# responsive-classes
A Vanilla JS library that adds classes to the `<body>` which can be used to apply responsive styles.
It is lightweight (673B minified and gzipped), has no dependencies and is documented with comments. Check out `responsive-classes.js` file.

## Dafault breakpoints:
| Class name | Maximum width  |
|--|--|
|.xs | 544px |
|.sm | 767px |
|.md | 992px |
|.lg | 1199px |
|.xl | 10000px |

## Usage
### Import
````js
import * as responsiveClasses from 'responsive-classes';
````
or
````html
<script src="./path/to/responsive-classes.min.js"></script>
````
### Default usage
To use default classes simply call the function
````js
responsiveClasses();
````
### Custom classes
To use custom classes provide an argument to the `responsiveClasses()` function like this:
````js
responsiveClasses({
	sm: 767,
	md: 992,
	lg: 1199,
	xl: 3000,
	xxl: 10000
});
````
The argument should be an object where property name is a class name and value is the maximum `<body>` width in pixels. Values should be a Number. The order is not important.

Body width used for calculations doesn't include scrollbar.

### Helper classes
Helper classes allow you to style multiple breakpoints using one class. Two helper classes are created for each breakpoint. One with suffix `-up` is created for screen width bigger than minimun width of a breakpoint, and class with suffix `-down` is created for screen width smaller than maximum width of a breakpoint.

To use helper classes provide `true` as a second argument to `responsiveClasses()`
````js
responsiveClasses(breakpoints, true);
````
You have to provide breakpoints object as the first argument if you want to use helper classes.

Assuming that you use three breakpoint like this:
````js
responsiveClasses({
	sm: 767,
	md: 992,
	lg: 1199
}, true);
````
You will get the following classes:

| Example screen width | Classes  |
|--|--|
| 640px | .sm .sm-down .sm-up .md-down .lg-down |
| 1024px | .md .md-down .md-up .sm-up .lg-down |
| 1600px | .lg .lg-down .lg-up .md-up .sm-up |

Screen width values in the table above are used only for presentation

## Browser support
responsive-classes should work on all modern web browsers including IE11 and mobile browsers
Tested on:
- IE11
- Safari 11
- New Edge
- New Firefox
- New Chrome
- Mobile Chrome
- Mobile Safari

## About
I was inspired by this [article](https://medium.com/@kelin2025/writing-js-libraries-less-than-1tb-size-6342da0c006a) on writing nano libraries in JavaScript. The code is written in Vanilla JS and has no dependencies. It's not only lightweight (673B minified and gzipped), but it also has a clean interface, extensive validation and is well documented with comments.

## License
MIT


CHANGELOG.md
# Changelog
All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](http://keepachangelog.com/en/1.0.0/)
and this project adheres to [Semantic Versioning](http://semver.org/spec/v2.0.0.html).

## [0.3.0] - 2018-10-02 "Shaked"

### Added
- Add additional classes to the body in the demo
- Add minified version of the library to demo directory, so it can be easily tested
- Add browser support section to README

### Changed
- Replace find method with forEach to support IE11
- Replace forEach method with for loop on element classList to support IE11
- Slightly improve demo

### Fixed
- Prevent addClasses method from deleting all classes from the body

## [0.2.0] - 2018-09-13 "Pikon"

### Added
- Helper classes
- Encapsulate library code
- Export as AMD module
- Allow direct script import in HTML
- Add comments to the code
- UglifyJS to minify and compress the library
- NPM script to run UglifyJS
- Add minified version
- New sections added to readme

### Fixed
- Fix throttling resize event

### Changed
- Refactor code
- Change main file name to `responsive-classes.js`
- Improve demo


## [0.1.0] - 2018-08-21

### Added
- Export as a Node.js module
- Throttle resize event and set it to 15 times per second
- Add README
- Add .gitignore

### Changed
- Change default classes
- Body width is used for calculations instead of window width

### Fixed
- Change ES6 spread operator to ES5 forEach method

## [0.0.1] - 2018-08-08
### Added
- Option to add one responsive class at the time to the body
- Arguments validation

demo/index.js
window.onload = function() {
  var config = {
    xs: 480,
    sm: 767,
    md: 992,
    lg: 1199,
    xl: 10000
  };
  responsiveClasses(config, true);

  var currentClasses = document.getElementById('current-classes');
  displayCurrentClasses();

  window.onresize = function() {
    displayCurrentClasses();
  };

  setTimeout(function() {
    document.body.classList.add('added-later');
  }, 4000);

  function formatCurrentClasses() {
    var classNames = document.body.className.split(' ');
    classNames.splice(classNames.indexOf('initial'), 1);
    var addedLater = classNames.indexOf('added-later');
    if (addedLater > -1) {
      classNames.splice(addedLater, 1);
    }
    return classNames.join(' ');
  }

  function displayCurrentClasses() {
    currentClasses.innerHTML = formatCurrentClasses();
  }
};


responsive-classes.js
(function(window, exportName){
  // breakpoint which are used if user defined breakpoints are not specified
  var breakpoints = {
    xs: 544,
    sm: 767,
    md: 992,
    lg: 1199,
    xl: 10000
  };
  var availableClasses = [];
  var resizeTimeout;

  // main function which is exported
  function responsiveClasses(configBreakpoints, helperClasses) {
    if (configBreakpoints) {
      // Validate arguments
      validateBreakpoints(configBreakpoints);
      checkHelperClassesValueType(helperClasses);
      breakpoints = configBreakpoints;
    }

    // Formats breakpoints to a collection that is used later on
    formatBreakpoints();

    // Fires function that add classes for the first time
    addClasses(document.body.clientWidth, helperClasses);
    // Update classes on resize with throttling, 15 times per second
    window.addEventListener("resize", function() {
      if ( !resizeTimeout ) {
        resizeTimeout = setTimeout(function() {
          addClasses(document.body.clientWidth, helperClasses);
          resizeTimeout = null;
        }, 66);
      }
    });
  }

  function formatBreakpoints() {
    // values array will consist of breakpoints values, 0 is for the minWidth of first (smallest) breakpoint
    var values = [0];
    var formatted = [];

    // Get values and class names of all breakpoints
    for (var breakpoint in breakpoints) {
      if (breakpoints.hasOwnProperty(breakpoint)) {
        values.push(breakpoints[breakpoint]);
        availableClasses.push(breakpoint);
      }
    }

    // Sort the values to correctly specify minWidth for each breakpoint later on
    values.sort(function(a, b) {
      return a - b;
    });

    // Create an object for each breakpoint with name, maxWidth and minWidth, and store them in termporaty formatted array
    for (var breakpoint in breakpoints) {
      if (breakpoints.hasOwnProperty(breakpoint)) {
        var maxWidth = breakpoints[breakpoint];
        formatted.push({
          name: breakpoint,
          maxWidth: maxWidth,
          minWidth: values[values.indexOf(maxWidth) - 1]
        });
      }
    }

    // Sort breakpoints based on maxWidth value, so they will be in order needed later on
    formatted.sort(function (a, b) {
      return a.maxWidth - b.maxWidth;
    });

    breakpoints = formatted; 
  }

  function addClasses(windowWidth, helperClasses) {
    var className;
    // cannot use find method because IE11 doesn't support it
    breakpoints.forEach(function(breakpoint) {
      if (breakpoint.minWidth < windowWidth && breakpoint.maxWidth >= windowWidth) {
        className = breakpoint.name;
      }
    });
    
    if (helperClasses) {
      // remove all helper classes currently added to the body
      for (var index = 0; index < document.body.classList.length; index++) {
        var currentClass = document.body.classList[index];
        if (currentClass.indexOf('-up') > -1|| currentClass.indexOf('-down') > -1) {
          document.body.classList.remove(currentClass);
        }
      }
      
      // prepers new helper classes and adds them to the body
      buildHelperClasses(breakpoints, windowWidth);
    }

    // IE support
    availableClasses.forEach(function(currentClass) {
      document.body.classList.remove(currentClass);
    });
    document.body.classList.add(className);
  }

  function buildHelperClasses(breakpoints, windowWidth) {
    var helperClasses = [];
    breakpoints.forEach(function(breakpoint) {
      if (breakpoint.minWidth < windowWidth) {
        helperClasses.push(breakpoint.name + '-up');
      }
      if (breakpoint.maxWidth >= windowWidth) {
        helperClasses.push(breakpoint.name + '-down');
      }
    });

    // IE support
    helperClasses.forEach(function(className) {
      document.body.classList.add(className);
    });
  }


  /**
   * Validation 
   */
  function validateBreakpoints(configBreakpoints) {
    checkBreakpointsType(configBreakpoints);
    checkIfUniqueBreakpointsProvided(configBreakpoints);
  }

  function checkBreakpointsType(configBreakpoints) {
    if (typeof configBreakpoints !== 'object' || configBreakpoints instanceof Array) {
      throw new TypeError("Config is not an object. Please provide an object");
    }
  }

  function checkIfUniqueBreakpointsProvided(configBreakpoints) {
    var values = [];

    for (var breakpoint in configBreakpoints) {
      if (configBreakpoints.hasOwnProperty(breakpoint)) {
        checkBreakpointValueType(configBreakpoints[breakpoint]);
        values.push(configBreakpoints[breakpoint]);
      } 
    }

    values.forEach(function(value) {
      var testAgainst = values.filter(function(item) {
        return item !== value;
      });

      if (testAgainst.length !== values.length - 1) {
        throw new Error("All breakpoints should have unique values");
      }
    });
  }

  function checkBreakpointValueType(breakpoint) {
    if (isNaN(breakpoint)) {
      throw new TypeError("All breakpoints should have numeric values");
    } else if (breakpoint < 1) {
      throw new Error("All breakpoints should have positive values");
    }
  }

  function checkHelperClassesValueType(helperClasses) {
    if (helperClasses !== undefined && helperClasses !== true && helperClasses !== false) {
      throw new TypeError("Second argument should be a boolean");
    }
  }

  /**
   *  Exports
   */
  if (typeof define === 'function' && define.amd) {
    define(function () {
      return responsiveClasses;
    });
  } else if (typeof module !== 'undefined' && module.exports) {
    module.exports = responsiveClasses;
  } else {
    window[exportName] = responsiveClasses;
  }
})(window, 'responsiveClasses');